### PR TITLE
always try inlining functions which do not call other functions

### PR DESCRIPTION
**Pull request by lcnr**
This is still a bit broken.

r? @wesleywiser

### CODE REVIEW

**Comment by tmiasko** [comment]
FYI: If the intention is to enable this on mir-opt-level=1, it might be necessary to resolve #71793 first.

**Comment by lcnr** [comment]
looks like ra also breaks because of this

File: library/core/src/iter/range.rs [hunk 1]
```diff
@@ -201,24 +201,12 @@ macro_rules! step_identical_methods {
...
         #[inline]
         fn forward(start: Self, n: usize) -> Self {
-            // In debug builds, trigger a panic on overflow.
-            // This should optimize completely out in release builds.
-            if Self::forward_checked(start, n).is_none() {
-                let _ = Add::add(Self::MAX, 1);
-            }
-            // Do wrapping math to allow e.g. `Step::forward(-128i8, 255)`.
-            start.wrapping_add(n as Self)
+            start + (n as Self)
```
Comment (cuviper) [inline library/core/src/iter/range.rs:204]: I think the intention here was to inherit the user's setting for debug assertion, even when `std` is compiled in release mode (as it usually is).

Comment (lcnr) [reply library/core/src/iter/range.rs:204]: yeah, will have to rewrite this again before merging :+1:

**Comment by leonardo-m** [comment]
I think "functions which do not call other function" used to be called leaf functions.

**Comment by lcnr** [comment]
closing this for now as there are still some bugs with inlining we have to fix first

**Comment by lcnr** [comment]
rebased on top of #77306 meaning that all inlining ICE I know of are fixed,

let's see if we can get a perf run going here

**Comment by lcnr** [comment]
@bors try @rust-timer queue

looks like all test failures are expected

**Comment by lcnr** [comment]
I first rebased and then opened

let's open another PR

**Comment by lcnr** [comment]
@Mark-Simulacrum should I do something to cancel rust-timer here?

**Comment by Mark-Simulacrum** [comment]
Don't worry about it -- it'll just stick around but that's not a problem.

**Pull request closed without merging**